src/module/data/abstract.ts

```diff
@@ -22,43 +18,48 @@ const immiscibleKeys = new Set([
+type ImmiscibleKeys = typeof immiscibleKeys extends Set<infer Key> ? Key : never
 
-// Combine instance types from templates
-type CombinedInstanceType<T extends Array<typeof SystemDataModel<any, any>>> = T extends [
-	infer U extends typeof SystemDataModel<any, any>,
-	...infer Rest extends Array<typeof SystemDataModel<any, any>>,
+declare class AnySystemDataModel extends SystemDataModel<any, any> {
+	constructor(...args: any[])
+}
+
+// Combined class type
+type Mixed<T extends Array<typeof AnySystemDataModel>> = ImmiscibleToConcrete<_Mixed<T>>
+
+// Combine static types, excluding immiscible keys
+type _Mixed<T extends (typeof AnySystemDataModel)[]> = T extends [
+	infer First extends typeof AnySystemDataModel,
+	...infer Rest extends (typeof AnySystemDataModel)[],
 ]
-	? InstanceType<U> & CombinedInstanceType<Rest>
+	? ImmiscibleToAny<First> & _Mixed<Rest>
 	: {}
 
-// Combine static types, excluding immiscible keys
-type CombinedStaticType<T extends any[]> = Omit<
-	T extends [infer U, ...infer Rest] ? U & CombinedStaticType<Rest> : {},
-	ImmiscibleKeys
->
+// @ts-expect-error - This is effectively a faux subclass of `T` which tsc isn't too fond of.
+// It can be unsound but in this case there's no other way to do it.
+interface ImmiscibleToAny<T extends object> extends Record<ImmiscibleKeys, any>, T {}
 
-// Combined class type
-type CombinedDataModel<
-	T extends Array<typeof SystemDataModel<any, any>>,
-	S extends DataSchema = MergeSchemas<T>,
-	B extends typeof SystemDataModel<any, any> = typeof SystemDataModel<any, any>,
-> = B &
-	CombinedStaticType<T> & {
-		new (...args: any[]): CombinedInstanceType<T> & InstanceType<B> & SystemDataModel<S>
-	}
-// type CombinedDataModel<
-// 	T extends Array<typeof SystemDataModel<any, any>>,
-// 	S extends DataSchema = MergeSchemas<T>,
-// > = CombinedStaticType<T> & {
-// 	new (...args: any[]): CombinedInstanceType<T> & SystemDataModel<S>
-// }
+// @ts-expect-error - This is effectively a faux subclass of `T` which tsc isn't too fond of.
+// It can be unsound but in this case there's no other way to do it.
+interface ImmiscibleToConcrete<T extends object> extends ConcreteImmiscible, T {}
 
-type MergeSchemas<T extends any[]> = {
-	[K in keyof T]: T[K] extends typeof SystemDataModel<infer S extends DataSchema, any> ? S : never
-}[number] extends infer U
-	? { [K in keyof U]: U[K] }
-	: never
+interface ConcreteImmiscible {
+	prototype: typeof AnySystemDataModel.prototype
+	length: number
+	name: string
+	cleanData: typeof AnySystemDataModel.cleanData
+	validateJoint: typeof SystemDataModel.validateJoint
+	migrateData: typeof SystemDataModel.migrateData
+	shimData: typeof SystemDataModel.shimData
+	defineSchema: typeof SystemDataModel.defineSchema
+
+	mixed: unknown // This may be relevant to your repo.
+	// These are probably dnd5e constructs.
+	_cleanData: unknown
+	_validateJoint: unknown
+	_migrateData: unknown
+	_shimData: unknown
```

Comment:
I didn't know how to fill these in, I assume at least some of these come from DND5e.